share/wake/lib/core/list.wake

```diff
@@ -179,63 +268,99 @@ export def scanr (combiningFn: element => accumulator => accumulator): accumulat
 # Concatenate two lists into one List.
 # The ++ operator must be provided by the wake package for publish to work.
 #
+# Guarantees:
+#  - ++ is associtive
+#  - Nil is an identity of ++
+#
+# Examples:
+#   ```
 #   (1, 2, 3, Nil) ++ (8, 9, Nil) = (1, 2, 3, 8, 9, Nil)
+#   Nil ++ x = x
+#   x ++ Nil = x
+#   ```
 export def (l: List a) ++ (r: List a): List a = lcat r l
 # curried to expose recursive loop outside of ++
 def lcat = foldr (_,_)
 
 # reverse: reverse the order of elements in a List
 #
+# Guarantees:
+#  - reverse is its own inverse
+#  - The resultant List has the same length as `list`
+#  - The resultant List is a permutation of `list`
+#
+# Examples:
+#   ```
 #   reverse (seq 5) = 4, 3, 2, 1, 0, Nil
 #   reverse Nil = Nil
+#   reverse (reverse x) = x
+#   ```
 export def reverse: List a => List a =
     foldl (\a\b b,a) Nil
 
 # flatten: concatenate a List of Lists into a simple List
 #
-#   flatten ((1, 2, Nil), (3, Nil), (4, 5, Nil), Nil)
-#   = 1, 2, 3, 4, 5, Nil
+# Examples:
+#   ```
+#   flatten ((1, 2, Nil), (3, Nil), (4, 5, Nil), Nil) = 1, 2, 3, 4, 5, Nil
+#   flatten Nil = Nil
+#   flatten (x, Nil) = x
+#   flatten (x, y, Nil) = x ++ y
+#   ```
 export def flatten: List (List a) => List a =
     foldr (_++_) Nil
 
 # len: count the number of elements in a List
 #
+# Examples:
+#   ```
 #   len Nil = 0
 #   len (1, 2, Nil) = 2
 #   len (seq x) = x
+#   ```
 export def len: List a => Integer =
     foldl (\a\_ a+1) 0
 
 # prepend: add `item` to the start of `list`.
 # This is efficient no matter how large the list.
 # Consider using ',' directly as it is more idiomatic.
 #
+# Examples:
+#   ```
 #   prepend 9 (seq 3) = 9, 0, 1, 2, Nil
 #   prepend 8 Nil     = 8, Nil
+#   ```
 export def prepend (item: a) (list: List a): List a =
     item, list
 
 # append: add `item` to the end of `list`.
-# This function must recreate the entire list to add the item.
+#
+# **NOTE**: This function must recreate the entire list to add the item.
```

Comment:
It's probably something to do in a separate PR, but I wonder if we want to add these asterisks to the "Examples" headers as well, so they're bolded in the popups.  I'm also wondering whether it's better to have the colon inside or outside the formatting.

Reply:
Yeah good question. I vote we get the content of this PR into shape and make sure its at least consistent with itself and then we can make sweeping changes like that after a good-ol'-fashioned bikeshed later.